plasmapy/diagnostics/thomson.py

```diff
@@ -0,0 +1,187 @@
+        scattering, while alpha < 1 indicates non-collective scattering.
+
+    Skw : astropy.units.Quantity ndarray
+        Computed spectral density function over the input wavelength range
+        with units of nm.
+
+    Notes
+    -----
+    For details, see "Plasma Scattering of Electromagnetic Radiation" by
+    Sheffield et al. ISBN 978-0123748775. This code is a modified version of
+    the program described therein.
+
+    For a concise summary of the relevant physics, see Chapter 5 of Derek
+    Schaeffer's thesis, DOI: 10.5281/zenodo.3766933
+
+    """
+
+    if np.sum(fract) != 1.0:
+        print("WARNING: Sum(fract) != 1.0. Normalizing array.")
+        fract = fract/np.sum(fract)
+
+    # Ensure unit vectors are normalized
+    probe_n = probe_n/np.linalg.norm(probe_n)
+    scatter_n = scatter_n/np.linalg.norm(scatter_n)
+
+    # Convert units
+    wavelength = wavelength.to(u.cm)
+    probe_wavelength = probe_wavelength.to(u.cm)
+    ne = ne.to(u.cm**-3)
+    Te = Te.to(u.eV)
+    Ti = Ti.to(u.eV)
+    ion_vel = ion_vel.to(u.cm/u.s)
+    fluid_vel = fluid_vel.to(u.cm/u.s)
+
+    # Define some constants
+    C = 2.99792458e10*u.cm/u.s  # speed of light
+    Me = 5.685e-16*u.eV/(u.cm/u.s)**2  # Electron mass in eV per (cm/s)^2
+    Mp = Me*1836.1  # Proton mass in same units
+    Mi = ion_mu*Mp  # mass of each ion species in same units
+    vTe = np.sqrt(Te/Me)  # Electron thermal velocity
+    vTi = np.sqrt(Ti/Mi)  # Ion thermal velocity
+    wpe = 5.64e4*(u.cm**1.5*u.rad/u.s)*np.sqrt(ne)
+
+    # Compute the ion velocity in the rest frame
+    ion_vel = fluid_vel + ion_vel
+
+    # Convert wavelengths to angular frequencies (electromagnetic waves, so
+    # phase speed is c)
+    ws = (2*np.pi*u.rad*C/wavelength).to(u.rad/u.s)
+    wl = (2*np.pi*u.rad*C/probe_wavelength).to(u.rad/u.s)
+
+    # Compute the frequency shift (required by energy conservation)
+    w = ws - wl
+
+    # Compute the wavenumbers in the plasma
+    # See Sheffield Sec. 1.8.1 and Eqs. 5.4.1 and 5.4.2
+    ks = np.sqrt(ws**2 - wpe**2)/C
+    kl = np.sqrt(wl**2 - wpe**2)/C
+
+    # Compute the wavenumber shift (required by momentum conservation)
+    sa = np.arccos(np.dot(probe_n, scatter_n))  # The scattering angle
+    k = np.sqrt(ks**2 + kl**2 - 2*ks*kl*np.cos(sa))  # Eq. 1.7.10 in Sheffield
+    k_n = scatter_n - probe_n  # Normal vector along k
+
+    # Compute Doppler-shifted frequencies for both the ions and electrons
+    # Matmul is simultaneously conducting dot product over all wavelengths
+    # and ion components
+    w_e = w - k*np.dot(fluid_vel, k_n)
+    w_i = w - np.matmul(ion_vel, np.outer(k, k_n).T)
+
+    # Compute the scattering parameter alpha
+    # expressed here using the fact that v_th/w_p = root(2) * Debye length
+    alpha = wpe/(np.sqrt(2)*k*vTe)
+
+    # Calculate the normalized phase velocities (Sec. 3.4.2 in Sheffield)
+    xe = w_e/(k*np.sqrt(2)*vTe)
+    xi = 1/np.sqrt(2)*np.outer(1/vTi, 1/k)*w_i
```

Comment:
I'd kill for cleaner variable naming here :smile: no good ideas yet, though.

Reply:
I see your point: if it makes it any better x_e and x_i are the conventional notation in the literature (including the Sheffield book I cite). I can't think of a more concise way to say "phase velocities normalized to thermal velocities"

Reply:
... `normalized_phase_ve`... nah. Okay, I give up.